src/roi/Roi.ts

```diff
@@ -91,4 +138,516 @@ export class Roi {
+      feretDiameterMax: this.feretDiameters.maxDiameter,
+      aspectRatio: this.feretDiameters.aspectRatio,
+      fillRatio: this.fillRatio,
+      sphericity: this.sphericity,
+      roundness: this.roundness,
+      solidity: this.solidity,
+      perimeter: this.perimeter,
+      perimeterInfo: this.perimeterInfo,
+      externalIDs: this.externalIDs,
+      externalLengths: this.externalLengths,
+      internalIDs: this.internalIDs,
+      borderIDs: this.borderIDs,
+      external: this.external,
+      borderLengths: this.borderLengths,
+      boxIDs: this.boxIDs,
+    };
+  }
+}
+/**
+ *
+ * @param roi -ROI
+ * @returns object which tells how many pixels are exposed externally to how many sides
+ */
+function getPerimeterInfo(roi: Roi) {
+  let roiMap = roi.getMap();
+  let data = roiMap.data;
+  let one = 0;
+  let two = 0;
+  let three = 0;
+  let four = 0;
+
+  for (let x = 0; x < roiMap.width; x++) {
+    for (let y = 0; y < roiMap.height; y++) {
+      let target = (y + roi.minY) * roiMap.width + x + roi.minX;
+      if (data[target] === roi.id) {
+        let nbAround = 0;
+        if (x === 0) {
+          nbAround++;
+        } else if (roi.externalIDs.includes(data[target - 1])) {
+          nbAround++;
+        }
+
+        if (x === roiMap.width - 1) {
+          nbAround++;
+        } else if (roi.externalIDs.includes(data[target + 1])) {
+          nbAround++;
+        }
+
+        if (y === 0) {
+          nbAround++;
+        } else if (roi.externalIDs.includes(data[target - roiMap.width])) {
+          nbAround++;
+        }
+
+        if (y === roiMap.height - 1) {
+          nbAround++;
+        } else if (roi.externalIDs.includes(data[target + roiMap.width])) {
+          nbAround++;
+        }
+        switch (nbAround) {
+          case 1:
+            one++;
+            break;
+          case 2:
+            two++;
+            break;
+          case 3:
+            three++;
+            break;
+          case 4:
+            four++;
+            break;
+          default:
+        }
+      }
+    }
+  }
+  return { one, two, three, four };
+}
+
+function getExternal(roi: Roi) {
+  let total = 0;
+  let roiMap = roi.getMap();
+  let data = roiMap.data;
+
+  for (let x = 0; x < roi.width; x++) {
+    for (let y = 0; y < roi.height; y++) {
+      let target = (y + roi.minY) * roiMap.width + x + roi.minX;
+      if (data[target] === roi.id) {
+        // if a point around is not roi.id it is a border
+        if (
+          roi.externalIDs.includes(data[target - 1]) ||
+          roi.externalIDs.includes(data[target + 1]) ||
+          roi.externalIDs.includes(data[target - roiMap.width]) ||
+          roi.externalIDs.includes(data[target + roiMap.width])
+        ) {
+          total++;
+        }
+      }
+    }
+  }
+  return total + roi.box;
+}
+
+/**
+ *
+ * @param roi - ROI
+ * @returns the surface of holes in ROI
+ */
+function getHolesInfo(roi: Roi) {
+  let surface = 0;
+  let width = roi.getMap().width;
```

Comment:
If the value will not change we should take the habbit to define it as `const`. This applies everywhere.